File: src/merkle_trees/treeUtils.js
const poseidon = require("../../circomlib/src/poseidon.js");
const { BigNumber: BN } = require("ethers");

module.exports = {
  rootFromLeafAndPath(leaf, idx, merkle_path) {
    if (merkle_path.length > 0) {
      const depth = merkle_path.length;
      const merkle_path_pos = module.exports.idxToBinaryPos(idx, depth);
      var root = new Array(depth);
      let left =
        BN.from(leaf).toBigInt() -
        BN.from(merkle_path_pos[0]).toBigInt() *
          (BN.from(leaf).toBigInt() - BN.from(merkle_path[0]).toBigInt());
      let right =
        BN.from(merkle_path[0]).toBigInt() -
        BN.from(merkle_path_pos[0]).toBigInt() *
          (BN.from(merkle_path[0]).toBigInt() - BN.from(leaf).toBigInt());
      root[0] = poseidon([left, right]);
      for (var i = 1; i < depth; i++) {
        left =
          BN.from(root[i - 1]).toBigInt() -
          BN.from(merkle_path_pos[i]).toBigInt() *
            (root[i - 1] - BN.from(merkle_path[i]).toBigInt());
        right =
          BN.from(merkle_path[i]).toBigInt() -
          BN.from(merkle_path_pos[i]).toBigInt() *
            (BN.from(merkle_path[i]).toBigInt() - root[i - 1]);
        root[i] = poseidon([left, right]);
      }
      return root[depth - 1];
    } else {
      return leaf;
    }
  },

  innerNodesFromLeafAndPath(leaf, idx, merkle_path) {
    if (merkle_path.length > 0) {
      const depth = merkle_path.length;
      const merkle_path_pos = module.exports.idxToBinaryPos(idx, depth);
      var innerNodes = new Array(depth);
      let left =
        BN.from(leaf).toBigInt() -
        BN.from(merkle_path_pos[0]).toBigInt() *
          (BN.from(leaf).toBigInt() - BN.from(merkle_path[0]).toBigInt());
      let right =
        BN.from(merkle_path[0]).toBigInt() -
        BN.from(merkle_path_pos[0]).toBigInt() *
          (BN.from(merkle_path[0]).toBigInt() - BN.from(leaf).toBigInt());
      innerNodes[0] = poseidon([left, right]);
      for (var i = 1; i < depth; i++) {
        left =
          innerNodes[i - 1] -
          BN.from(merkle_path_pos[i]).toBigInt() *
            (innerNodes[i - 1] - BN.from(merkle_path[i]).toBigInt());
        right =
          BN.from(merkle_path[i]).toBigInt() -
          BN.from(merkle_path_pos[i]).toBigInt() *
            (BN.from(merkle_path[i]).toBigInt() - innerNodes[i - 1]);
        innerNodes[i] = poseidon([left, right]);
      }
      return innerNodes;
    } else {
      return leaf;
    }
  },

  proofPos: function (leafIdx, treeDepth) {
    let proofPos = new Array(treeDepth);
    let proofBinaryPos = module.exports.idxToBinaryPos(leafIdx, treeDepth);

    if (leafIdx % 2 == 0) {
      proofPos[0] = leafIdx + 1;
    } else {
      proofPos[0] = leafIdx - 1;
    }

    for (var i = 1; i < treeDepth; i++) {
      if (proofBinaryPos[i] == 1) {
        proofPos[i] = Math.floor(proofPos[i - 1] / 2) - 1;
      } else {
        proofPos[i] = Math.floor(proofPos[i - 1] / 2) + 1;
      }
    }

    return proofPos;
  },

  getAffectedPos: function (proofPos) {
    var affectedPos = new Array(proofPos.length);

    // skip the first node in the proof since it is not affected
    for (var i = 1; i < proofPos.length; i++) {
      // if proof node has odd index (i.e. is the right sibling)
      if (proofPos[i] & 1) {
        affectedPos[i - 1] = proofPos[i] - 1; // affected node is left sibling
        // if proof node has even index (i.e. is the left sibling)
      } else {
        affectedPos[i - 1] = proofPos[i] + 1; // affected node is right sibling
      }
    }

    affectedPos[proofPos.length - 1] = 0; // the root

    return affectedPos;
  },

  binaryPosToIdx: function (binaryPos) {
    var idx = 0;
    for (i = 0; i < binaryPos.length; i++) {
      idx = idx + binaryPos[i] * 2 ** i;
    }
    return idx;
  },

  idxToBinaryPos: function (idx, binLength) {
    let binString = idx.toString(2);
    let binPos = Array(binLength).fill(0);
    for (var j = 0; j < binString.length; j++) {
      binPos[j] = Number(binString.charAt(binString.length - j - 1));
    }
    return binPos;
  },

  pairwiseHash: function (array) {
    if (array.length % 2 == 0) {
      let arrayHash = [];
      for (var i = 0; i < array.length; i = i + 2) {
        arrayHash.push(
          poseidon([array[i].toString(), array[i + 1].toString()])
        );
      }
      return arrayHash;
    } else {
      console.log("array must have even number of elements");
    }
  },

  getBase2Log: function (y) {
    return Math.log(y) / Math.log(2);
  },

  // fill an array with a fillerLength copies of a value
  padArray: function (leafArray, padValue) {
    if (Array.isArray(leafArray)) {
      var arrayClone = leafArray.slice(0);
      const nearestPowerOfTwo = Math.ceil(
        module.exports.getBase2Log(leafArray.length)
      );
      const diff = 2 ** nearestPowerOfTwo - leafArray.length;
      for (var i = 0; i < diff; i++) {
        arrayClone.push(padValue);
      }
      return arrayClone;
    } else {
      console.log("please enter an array");
    }
  },
};

File: invisible_backend/src/trees/tree_utils.rs
use num_bigint::BigUint;
use std::{fmt::Binary, str::FromStr};

use crate::pedersen::pedersen;

pub fn pairwise_hash(array: &Vec<BigUint>) -> Vec<BigUint> {
    if array.len() % 2 != 0 {
        panic!("Array length must be even");
    }

    let mut hashes: Vec<BigUint> = Vec::new();
    for i in (0..array.len()).step_by(2) {
        let hash: BigUint = pedersen(&array[i], &array[i + 1]);
        hashes.push(hash);
    }

    return hashes;
}

pub fn idx_to_binary_pos(idx: u64, bin_length: usize) -> Vec<i8> {
    // bin_length = depth

    let bin_chars = format!("{idx:b}");

    assert!(
        bin_chars.len() <= bin_length,
        "index is to big to fit on the tree"
    );

    let mut bin_pos: Vec<i8> = Vec::new();

    for ch in bin_chars.chars() {
        // println!("{:?}", ch);
        bin_pos.push(ch.to_digit(10).unwrap() as i8)
    }

    for _ in 0..bin_length - bin_chars.len() {
        bin_pos.insert(0, 0);
    }

    bin_pos.reverse();

    return bin_pos;
}

pub fn proof_pos(leaf_idx: u64, depth: usize) -> Vec<u64> {
    let mut proof_pos: Vec<u64> = Vec::new();
    let proof_binary_pos = idx_to_binary_pos(leaf_idx, depth);

    if leaf_idx % 2 == 0 {
        proof_pos.push(leaf_idx + 1);
    } else {
        proof_pos.push(leaf_idx - 1);
    }

    for i in 1..depth {
        if proof_binary_pos[i] == 1 {
            let pos_i = proof_pos[i - 1] / 2 - 1;
            proof_pos.push(pos_i);
        } else {
            let pos_i = proof_pos[i - 1] / 2 + 1;
            proof_pos.push(pos_i);
        }
    }

    return proof_pos;
}

pub fn get_affected_pos(proof_pos: Vec<u64>) -> Vec<u64> {
    let mut affected_pos: Vec<u64> = Vec::new();

    for i in 1..proof_pos.len() {
        if proof_pos[i] % 2 == 1 {
            affected_pos.push(proof_pos[i] - 1);
        } else {
            affected_pos.push(proof_pos[i] + 1);
        }
    }

    affected_pos.push(0);

    return affected_pos;
}

pub fn inner_nodes_from_leaf_and_path(
    leaf_hash: &BigUint,
    leaf_idx: u64,
    merkle_path: &Vec<BigUint>,
) -> Vec<BigUint> {
    let depth = merkle_path.len();
    let merkle_path_pos = idx_to_binary_pos(leaf_idx, depth);

    let mut inner_nodes: Vec<BigUint> = Vec::new();

    // Left
    let left: BigUint;
    let right: BigUint;
    if merkle_path_pos[0] == 0 {
        left = leaf_hash.clone();
        right = merkle_path[0].clone();
    } else {
        left = merkle_path[0].clone();
        right = leaf_hash.clone();
    };

    inner_nodes.push(pedersen(&left, &right));

    for i in 1..depth {
        // Left
        let left: BigUint;
        let right: BigUint;
        if merkle_path_pos[i] == 0 {
            left = inner_nodes[i - 1].clone();
            right = merkle_path[i].clone();
        } else {
            left = merkle_path[i].clone();
            right = inner_nodes[i - 1].clone();
        };

        let hash: BigUint = pedersen(&left, &right);
        inner_nodes.push(hash);
    }

    return inner_nodes;
}
